src/Microsoft.DotNet.Interactive.ExtensionLab/MsSqlKernel.cs

```diff
@@ -3,21 +3,155 @@
+        private bool _intellisenseReady = false;
+        private readonly string _tempFilePath;
+        private readonly string _tempFileUri;
         private readonly string _connectionString;
+        private readonly MsSqlServiceClient _serviceClient;
+
+        private TaskCompletionSource<ConnectionCompleteParams> _connectionCompleted = new TaskCompletionSource<ConnectionCompleteParams>();
+
+        private ConcurrentDictionary<string, Action<QueryCompleteParams>> _queryHandlers = new ConcurrentDictionary<string, Action<QueryCompleteParams>>();
 
         public MsSqlKernel(string name, string connectionString) : base(name)
         {
+            _tempFilePath = Path.GetTempFileName();
+            _tempFileUri = MsSqlServiceClient.GetUriForFilePath(_tempFilePath);
             _connectionString = connectionString;
+            _serviceClient = new MsSqlServiceClient();
+
+            _serviceClient.OnConnectionComplete += HandleConnectionComplete;
+            _serviceClient.OnQueryComplete += HandleQueryComplete;
+            _serviceClient.OnIntellisenseReady += HandleIntellisenseReady;
+
+            _serviceClient.StartProcessAndRedirectIO();
+
+            RegisterForDisposal(_serviceClient);
+        }
+
+        private void HandleConnectionComplete(object sender, ConnectionCompleteParams connParams)
+        {
+            if (connParams.OwnerUri.Equals(_tempFileUri))
+            {
+                if (connParams.ErrorMessage != null)
+                {
+                    _connectionCompleted.SetException(new Exception(connParams.ErrorMessage));
+                }
+                else
+                {
+                    _connectionCompleted.SetResult(connParams);
+                }
+            }
+        }
+
+        private void HandleQueryComplete(object sender, QueryCompleteParams queryParams)
+        {
+            Action<QueryCompleteParams> handler;
+            if (_queryHandlers.TryGetValue(queryParams.OwnerUri, out handler))
+            {
+                handler(queryParams);
+            }
+        }
+
+        private void HandleIntellisenseReady(object sender, IntelliSenseReadyParams readyParams)
+        {
+            if (readyParams.OwnerUri.Equals(this._tempFileUri))
+            {
+                this._intellisenseReady = true;
+            }
         }
 
-        protected override DbConnection OpenConnection()
+        public async Task HandleAsync(SubmitCode command, KernelInvocationContext context)
         {
-            return new SqlConnection(_connectionString);
+            if (!_connected)
+            {
+                await _serviceClient.ConnectAsync(_tempFileUri, _connectionString);
+                await _connectionCompleted.Task;
+                _connected = true;
+            }
+
+            var queryResult = await _serviceClient.ExecuteSimpleQueryAsync(_tempFileUri, command.Code);
+            var tableString = GetTableStringForSimpleResult(queryResult);
```

Comment:
In general we want kernels to focus on returning data and not worry about MIME types or formatting. This will make them more flexible. The `SqliteKernel` approach might be a good shape, or we can come up with something different, but ideally it will be consistent across data kernels.